### ISSUE TITLE
Error while running ra-tls-mbedtls sample

### ISSUE THREAD

**Issue by IceCY** [gramineproject/gramine#1350]
### Description of the problem

I am using Gramine v1.4. I want to perform "RA-TLS flows with SGX and with Gramine, running DCAP client in SGX" for ra-tls-mbedtls sample. I follow the guidence and compile the sample as below.

``` Bash
make clean
make app dcap RA_TYPE=dcap
```

During running, I got the following error information
```
  . Seeding the random number generator... ok
  . Connecting to tcp/localhost/4433... ok
  . Setting up the SSL/TLS structure... ok
  . Loading the CA root certificate ... ok
  . Installing RA-TLS callback ... ok
  . Performing the SSL/TLS handshake...
[get_driver_type edmm_utility.cpp:116] Failed to open Intel SGX device.
[open_se_device edmm_utility.cpp:178] Failed to open Intel SGX device.
[get_driver_type /home/icecy/sgx/linux-sgx/psw/urts/linux/edmm_utility.cpp:116] Failed to open Intel SGX device.
[get_driver_type edmm_utility.cpp:116] Failed to open Intel SGX device.
[open_se_device edmm_utility.cpp:178] Failed to open Intel SGX device.
[get_driver_type /home/icecy/sgx/linux-sgx/psw/urts/linux/edmm_utility.cpp:116] Failed to open Intel SGX device.
ra_tls_verify_callback: sgx_qv_verify_quote failed: 57372
 failed
  ! mbedtls_ssl_handshake returned -0x3000
```

It seems like that the program is trying to load and create QvE enclave inside the enclave. Because I get the following output for `strace`. 

```
open("/lib/x86_64-linux-gnu/libsgx_urts.so.2", O_RDONLY|O_CLOEXEC) = 23
fstat(23, {st_mode=S_IFREG|0644, st_size=148744, ...}) = 0
mmap(NULL, 148744, PROT_READ, MAP_SHARED, 23, 0) = 0x7f9d1bbc9000
open("/lib/x86_64-linux-gnu/libsgx_urts.so.2", O_RDONLY|O_CLOEXEC) = 24
fstat(24, {st_mode=S_IFREG|0644, st_size=148744, ...}) = 0
mmap(NULL, 148744, PROT_READ, MAP_SHARED, 24, 0) = 0x7f9d1bba4000
munmap(0x7f9d1bba4000, 148744)          = 0
close(24)    
......
open("/lib/x86_64-linux-gnu/libsgx_qve.signed.so.1", O_RDONLY|O_NONBLOCK|O_CLOEXEC) = 25
fstat(25, {st_mode=S_IFREG|0644, st_size=3897240, ...}) = 0
close(25)                               = 0
open("/lib/x86_64-linux-gnu/libsgx_qve.signed.so.1", O_RDONLY|O_CLOEXEC) = 25
fstat(25, {st_mode=S_IFREG|0644, st_size=3897240, ...}) = 0
mmap(NULL, 3897240, PROT_READ, MAP_SHARED, 25, 0) = 0x7f9d1b805000
--- SIGILL {si_signo=SIGILL, si_code=ILL_ILLOPN, si_addr=0x55d0764154fc} ---
rt_sigreturn({mask=[USR2]})             = 3
--- SIGILL {si_signo=SIGILL, si_code=ILL_ILLOPN, si_addr=0x55d0764154fc} ---
rt_sigreturn({mask=[USR2]})             = 3
--- SIGILL {si_signo=SIGILL, si_code=ILL_ILLOPN, si_addr=0x55d0764154fc} ---
rt_sigreturn({mask=[USR2]})             = 3
write(1, "[get_driver_type edmm_utility.cp"..., 72) = 72
write(1, "[open_se_device edmm_utility.cpp"..., 71) = 71
```

I doubt maybe it is because that the `DUMMY_FUNCTION(sgx_create_enclave)` in `ra_tls_verify_dcap_gramine.c` is not invoked.

### Steps to reproduce

Follow the "RA-TLS flows with SGX and with Gramine, running DCAP client in SGX" part of ra-tls-mbedtls sample

### Expected results

ra-tls-mbedtls sample running successfully.

### Actual results

client verification failed with the following output.
```
  . Seeding the random number generator... ok
  . Connecting to tcp/localhost/4433... ok
  . Setting up the SSL/TLS structure... ok
  . Loading the CA root certificate ... ok
  . Installing RA-TLS callback ... ok
  . Performing the SSL/TLS handshake...
my_vefiry_callback
[get_driver_type edmm_utility.cpp:116] Failed to open Intel SGX device.
[open_se_device edmm_utility.cpp:178] Failed to open Intel SGX device.
[get_driver_type /home/icecy/sgx/linux-sgx/psw/urts/linux/edmm_utility.cpp:116] Failed to open Intel SGX device.
[get_driver_type edmm_utility.cpp:116] Failed to open Intel SGX device.
[open_se_device edmm_utility.cpp:178] Failed to open Intel SGX device.
[get_driver_type /home/icecy/sgx/linux-sgx/psw/urts/linux/edmm_utility.cpp:116] Failed to open Intel SGX device.
ra_tls_verify_callback: sgx_qv_verify_quote failed: 57372
 failed
  ! mbedtls_ssl_handshake returned -0x3000
```

### Gramine commit hash

aefd5d0875c4fdf76edd67f5043952bdd32b0f7a

**Comment by dimakuv** [comment]
> ra_tls_verify_callback: sgx_qv_verify_quote failed: 57372

This is the error code `hex(57372) = 0xe01c`. Which stands for `SGX_QL_QUOTE_CERTIFICATION_DATA_UNSUPPORTED`: https://github.com/intel/SGXDataCenterAttestationPrimitives/blob/4cb5c8b81f126f9aa3ee921d7980a909a9bd676d/QuoteGeneration/quote_wrapper/common/inc/sgx_ql_lib_common.h#LL79C5-L79C49

I don't think I've seen this error before. The official docs say this: `SGX_QL_QUOTE_CERTIFICATION_DATA_UNSUPPORTED:
The quote verifier doesn’t support the certification data in the Quote. Currently, the Intel QVE only supported CertType = 5.` See https://download.01.org/intel-sgx/sgx-dcap/1.3/linux/docs/Intel_SGX_ECDSA_QuoteLibReference_DCAP_API.pdf

> It seems like that the program is trying to load and create QvE enclave inside the enclave. 

That's true, but IIRC this is benign. The QL (Quoting Library) first tries to perform in-process verification for which it tries to create the QvE (Quote Verification Enclave). But then the QL should fall back to out-of-process verification for which it will try to communicate with the AESM daemon. (**UPDATE**: Ok, I think I remembered it wrong. Looks like QL does *not* fall back to out-of-process verification but just fails. Which this issue proves.)

But you can also skip the QvE attempt by QL, by specifying this environment variable: `SGX_AESM_ADDR=1`. Please check some resources:
- https://github.com/intel/SGXDataCenterAttestationPrimitives/issues/107
- https://learn.microsoft.com/en-us/azure/confidential-computing/confidential-nodes-aks-addon

**Comment by dimakuv** [comment]
> I tried `SGX_AESM_ADDR=1`. It seems that it does not work. I notice that you mentioned in another issue that `SGX_AESM_ADDR=1` is only for Quote Generation, rather than Quote Verification.

Hm, good point. So there is some other problem with your Quote Verification...

What else do you see in the `strace`? Maybe this will give us some hints.

**Comment by IceCY** [comment]
I compile the sample code in debug mode. The following is the output. The output of strace is also attached. Hope it's helpful. Thanks.
``` 
Gramine is starting. Parsing TOML manifest file, this may take some time...
(host_main.c:994:load_enclave) debug: Gramine parsed TOML manifest file successfully
(host_framework.c:232:create_enclave) debug: Enclave created:
(host_framework.c:233:create_enclave) debug:     base:           0x0000000020000000
(host_framework.c:234:create_enclave) debug:     size:           0x0000000020000000
(host_framework.c:235:create_enclave) debug:     misc_select:    0x00000000
(host_framework.c:236:create_enclave) debug:     attr.flags:     0x0000000000000007
(host_framework.c:237:create_enclave) debug:     attr.xfrm:      0x00000000000000e7
(host_framework.c:238:create_enclave) debug:     ssa_frame_size: 4
(host_framework.c:239:create_enclave) debug:     isv_prod_id:    0x00000000
(host_framework.c:240:create_enclave) debug:     isv_svn:        0x00000000
(host_main.c:487:initialize_enclave) debug: Adding pages to SGX enclave, this may take some time...
(host_framework.c:320:add_pages_to_enclave) debug: Adding pages to enclave: 0x3ff07000-0x40000000 [REG:R--] (manifest) measured
(host_framework.c:320:add_pages_to_enclave) debug: Adding pages to enclave: 0x3fee7000-0x3ff07000 [REG:RW-] (ssa) measured
(host_framework.c:320:add_pages_to_enclave) debug: Adding pages to enclave: 0x3fee3000-0x3fee7000 [TCS:---] (tcs) measured
(host_framework.c:320:add_pages_to_enclave) debug: Adding pages to enclave: 0x3fedf000-0x3fee3000 [REG:RW-] (tls) measured
(host_framework.c:320:add_pages_to_enclave) debug: Adding pages to enclave: 0x3fe9f000-0x3fedf000 [REG:RW-] (stack) measured
(host_framework.c:320:add_pages_to_enclave) debug: Adding pages to enclave: 0x3fe5f000-0x3fe9f000 [REG:RW-] (stack) measured
(host_framework.c:320:add_pages_to_enclave) debug: Adding pages to enclave: 0x3fe1f000-0x3fe5f000 [REG:RW-] (stack) measured
(host_framework.c:320:add_pages_to_enclave) debug: Adding pages to enclave: 0x3fddf000-0x3fe1f000 [REG:RW-] (stack) measured
(host_framework.c:320:add_pages_to_enclave) debug: Adding pages to enclave: 0x3fdcf000-0x3fddf000 [REG:RW-] (sig_stack) measured
(host_framework.c:320:add_pages_to_enclave) debug: Adding pages to enclave: 0x3fdbf000-0x3fdcf000 [REG:RW-] (sig_stack) measured
(host_framework.c:320:add_pages_to_enclave) debug: Adding pages to enclave: 0x3fdaf000-0x3fdbf000 [REG:RW-] (sig_stack) measured
(host_framework.c:320:add_pages_to_enclave) debug: Adding pages to enclave: 0x3fd9f000-0x3fdaf000 [REG:RW-] (sig_stack) measured
(host_framework.c:320:add_pages_to_enclave) debug: Adding pages to enclave: 0x3fd48000-0x3fd95000 [REG:R-X] (code) measured
(host_framework.c:320:add_pages_to_enclave) debug: Adding pages to enclave: 0x3fd95000-0x3fd9a000 [REG:RW-] (data) measured
(host_framework.c:320:add_pages_to_enclave) debug: Adding pages to enclave: 0x3fd9a000-0x3fd9f000 [REG:RW-] (bss) measured
(host_framework.c:320:add_pages_to_enclave) debug: Adding pages to enclave: 0x20000000-0x3fd48000 [REG:RWX] (free)
(host_main.c:569:initialize_enclave) debug: Added all pages to SGX enclave
(host_framework.c:523:init_enclave) debug: Enclave initializing:
(host_framework.c:524:init_enclave) debug:     enclave id:   0x000000003ffff000
(host_framework.c:525:init_enclave) debug:     mr_enclave:   210037a50aa65121da23a434b436ab311c21d684d9ec878316ea0f3316549acf
-----------------------------------------------------------------------------------------------------------------------
Gramine detected the following insecure configurations:

  - sgx.debug = true                           (this is a debug enclave)
  - loader.log_level = warning|debug|trace|all (verbose log level, may leak information)
  - loader.insecure__use_cmdline_argv = true   (forwarding command-line args from untrusted host to the app)
  - loader.insecure__use_host_env = true       (forwarding environment vars from untrusted host to the app)
  - sgx.allowed_files = [ ... ]                (some files are passed through from untrusted host without verification)

Gramine will continue application execution, but this configuration must not be used in production!
-----------------------------------------------------------------------------------------------------------------------

(libos_init.c:382:libos_init) debug: Gramine was built from commit: aefd5d0875c4fdf76edd67f5043952bdd32b0f7a
(libos_init.c:384:libos_init) debug: Host: Linux-SGX
(libos_context.c:70:libos_xstate_init) debug: LibOS xsave_enabled 1, xsave_size 0xa80(2688), xsave_features 0xe7
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fd37000-0x3fd48000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fd36000-0x3fd37000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fd35000-0x3fd36000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fd34000-0x3fd35000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fd33000-0x3fd34000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fd32000-0x3fd33000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fd31000-0x3fd32000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fd2f000-0x3fd31000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fd2e000-0x3fd2f000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fd2d000-0x3fd2e000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fd2b000-0x3fd2d000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fd28000-0x3fd2b000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fd26000-0x3fd28000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fd23000-0x3fd26000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fd21000-0x3fd23000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fd1c000-0x3fd21000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fd17000-0x3fd1c000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fd14000-0x3fd17000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fd12000-0x3fd14000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fd09000-0x3fd12000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fd00000-0x3fd09000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fcfa000-0x3fd00000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fcf5000-0x3fcfa000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fcf4000-0x3fcf5000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fce3000-0x3fcf4000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fcd2000-0x3fce3000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fcd0000-0x3fcd2000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fcc3000-0x3fcce000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fcc0000-0x3fcc3000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fc9d000-0x3fcbe000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fc76000-0x3fc97000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fc6d000-0x3fc76000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fc53000-0x3fc68000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fc42000-0x3fc53000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fc01000-0x3fc42000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fbd8000-0x3fc01000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fb97000-0x3fbd8000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fb66000-0x3fb97000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fb02000-0x3fb53000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fa81000-0x3fb02000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3fa60000-0x3fa81000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3f9cf000-0x3fa50000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3f92e000-0x3f9cf000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3f82d000-0x3f92e000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3f7ec000-0x3f82d000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3f76b000-0x3f7ec000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3f66a000-0x3f76b000 (PAL internal me) bookkeeped
(libos_vma.c:615:init_vma) debug: Initial VMA region 0x3f5b7000-0x3f66a000 (PAL internal me) bookkeeped
(libos_vma.c:646:init_vma) debug: ASLR top address adjusted to 0x2a093000
(attestation.c:410:init_sgx_attestation) debug: host is Linux-SGX and remote attestation type is 'none', skipping /dev/attestation/quote file
(libos_init.c:410:libos_init) debug: LibOS loaded at 0x3f5b7000, ready to initialize
(libos_init.c:568:create_pipe) debug: Creating pipe: pipe.srv:0823622220da911ed8dee6242006e12925f88d8b175dae92a8765be443bdc70b
(libos_fs.c:787:mount_fs) [P1:T1:] debug: mounting "file:." (chroot) under /
(libos_fs.c:787:mount_fs) [P1:T1:] debug: mounting "proc" (pseudo) under /proc
(libos_fs.c:787:mount_fs) [P1:T1:] debug: mounting "dev" (pseudo) under /dev
(libos_fs.c:787:mount_fs) [P1:T1:] debug: mounting "dev:tty" (chroot) under /dev/tty
(libos_fs.c:787:mount_fs) [P1:T1:] debug: mounting "sys" (pseudo) under /sys
(libos_fs.c:787:mount_fs) [P1:T1:] debug: mounting "file:/usr/lib/x86_64-linux-gnu/gramine/runtime/glibc" (chroot) under /lib
(libos_fs.c:787:mount_fs) [P1:T1:] debug: mounting "file:/usr/local/lib" (chroot) under /usr/local/lib
(libos_fs.c:787:mount_fs) [P1:T1:] debug: mounting "file:/lib/x86_64-linux-gnu" (chroot) under /lib/x86_64-linux-gnu
(libos_fs.c:787:mount_fs) [P1:T1:] debug: mounting "file:/usr/lib/x86_64-linux-gnu" (chroot) under /usr/lib/x86_64-linux-gnu
(libos_fs.c:787:mount_fs) [P1:T1:] debug: mounting "file:/etc" (chroot) under /etc
(libos_fs.c:787:mount_fs) [P1:T1:] debug: mounting "file:client" (chroot) under /client
(libos_fs.c:787:mount_fs) [P1:T1:] debug: mounting "emulated-etc-resolv-conf" (pseudo) under /etc/resolv.conf
(libos_init.c:568:create_pipe) [P1:T1:client] debug: Creating pipe: pipe.srv:6f58b4488f795327e373ece1f1157f4b14f906bbb55d65f6b75ee7282286f8a8
(libos_init.c:303:init_stack) [P1:T1:client] debug: Allocated stack at 0x2a053000 (size = 0x40000)
(libos_rtld.c:844:load_elf_object) [P1:T1:client] debug: loading "file:client"
(libos_rtld.c:901:find_interp) [P1:T1:client] debug: searching for interpreter: /lib/ld-linux-x86-64.so.2
(libos_rtld.c:844:load_elf_object) [P1:T1:client] debug: loading "file:/usr/lib/x86_64-linux-gnu/gramine/runtime/glibc/ld-linux-x86-64.so.2"
(libos_init.c:568:create_pipe) [P1:T1:client] debug: Creating pipe: pipe.srv:1
(libos_init.c:568:create_pipe) [P1:T1:client] debug: Creating pipe: pipe.srv:9bcd029a624c87f29bfeda05ad24b65535985108592ebbf102366f2d57a56db7
(libos_ipc_worker.c:372:ipc_worker_wrapper) [P1:libos] debug: IPC worker started
(libos_init.c:505:libos_init) [P1:T1:client] debug: LibOS initialized
(libos_rtld.c:1046:register_library) [P1:T1:client] debug: glibc register library /lib/libmbedcrypto_gramine.so.13 loaded at 0x29f84000
(libos_rtld.c:1046:register_library) [P1:T1:client] debug: glibc register library /lib/libmbedtls_gramine.so.19 loaded at 0x29f50000
(libos_rtld.c:1046:register_library) [P1:T1:client] debug: glibc register library /lib/libmbedx509_gramine.so.4 loaded at 0x29f3f000
(libos_rtld.c:1046:register_library) [P1:T1:client] debug: glibc register library /lib/libc.so.6 loaded at 0x29d6d000
(libos_parser.c:1598:warn_unsupported_syscall) [P1:T1:client] warning: Unsupported system call rseq
(libos_rtld.c:1046:register_library) [P1:T1:client] debug: glibc register library /lib/libra_tls_verify_dcap_gramine.so loaded at 0x29cde000
(libos_rtld.c:1046:register_library) [P1:T1:client] debug: glibc register library /lib/x86_64-linux-gnu/libsgx_dcap_quoteverify.so.1 loaded at 0x29abf000
(libos_rtld.c:1046:register_library) [P1:T1:client] debug: glibc register library /lib/x86_64-linux-gnu/libstdc++.so.6 loaded at 0x29895000
(libos_rtld.c:1046:register_library) [P1:T1:client] debug: glibc register library /lib/x86_64-linux-gnu/libgcc_s.so.1 loaded at 0x29875000
(libos_rtld.c:1046:register_library) [P1:T1:client] debug: glibc register library /lib/libm.so.6 loaded at 0x29797000
in_sgx: 1
libra_tls_verify_dcap_gramine is loaded...
[ using default SGX-measurement verification callback (via RA_TLS_* environment variables) ]

  . Seeding the random number generator... ok
  . Connecting to tcp/localhost/4433... ok
  . Setting up the SSL/TLS structure... ok
  . Loading the CA root certificate ... ok
  . Installing RA-TLS callback ... ok
  . Performing the SSL/TLS handshake...(pal_files.c:77:file_open) warning: Disallowing access to file '/etc/localtime'; file is not trusted or allowed.

my_vefiry_callback
(pal_files.c:77:file_open) warning: Disallowing access to file '/etc/ld.so.cache'; file is not trusted or allowed.
(libos_rtld.c:1046:register_library) [P1:T1:client] debug: glibc register library /lib/x86_64-linux-gnu/libsgx_urts.so.2 loaded at 0x2966e000
(libos_rtld.c:1046:register_library) [P1:T1:client] debug: glibc register library /lib/x86_64-linux-gnu/libsgx_enclave_common.so.1 loaded at 0x29661000
[get_driver_type edmm_utility.cpp:116] Failed to open Intel SGX device.
[open_se_device edmm_utility.cpp:178] Failed to open Intel SGX device.
[get_driver_type /home/icecy/sgx/linux-sgx/psw/urts/linux/edmm_utility.cpp:116] Failed to open Intel SGX device.
[get_driver_type edmm_utility.cpp:116] Failed to open Intel SGX device.
[open_se_device edmm_utility.cpp:178] Failed to open Intel SGX device.
[get_driver_type /home/icecy/sgx/linux-sgx/psw/urts/linux/edmm_utility.cpp:116] Failed to open Intel SGX device.
ra_tls_verify_callback: sgx_qv_verify_quote failed: 57372
 failed
  ! mbedtls_ssl_handshake returned -0x3000

Last error was: -12288 - X509 - A fatal error occurred, eg the chain is too long or the vrfy callback failed

(libos_exit.c:179:libos_syscall_exit_group) [P1:T1:client] debug: ---- exit_group (returning 1)
(libos_fs_lock.c:653:posix_lock_clear_pid) [P1:T1:client] debug: clearing POSIX locks for pid 1
(libos_sync_client.c:331:shutdown_sync_client) [P1:T1:client] debug: sync client shutdown: closing handles
(libos_sync_client.c:346:shutdown_sync_client) [P1:T1:client] debug: sync client shutdown: waiting for confirmation
(libos_sync_client.c:359:shutdown_sync_client) [P1:T1:client] debug: sync client shutdown: finished
(libos_ipc_worker.c:285:ipc_worker_main) [P1:libos] debug: IPC worker: exiting worker thread
(libos_exit.c:58:libos_clean_and_exit) [P1:T1:client] debug: process 1 exited with status 1
(pal_process.c:248:_PalProcessExit) debug: PalProcessExit: Returning exit code 1
```

[strace.txt](https://github.com/gramineproject/gramine/files/11488481/strace.txt)

**Comment by IceCY** [comment]
I also tried to run client outside the enclave following the sample guidence. The same error is produced. 
```
ra_tls_verify_callback: sgx_qv_verify_quote failed: 57372
 failed
  ! mbedtls_ssl_handshake returned -0x3000
```

**Comment by dimakuv** [comment]
> That's true, but IIRC this is benign. The QL (Quoting Library) first tries to perform in-process verification for which it tries to create the QvE (Quote Verification Enclave). But then the QL should fall back to out-of-process verification for which it will try to communicate with the AESM daemon. (UPDATE: Ok, I think I remembered it wrong. Looks like QL does not fall back to out-of-process verification but just fails. Which this issue proves.)

Replying to my own incorrect statement from before.

RA-TLS in Gramine actually **always** uses the non-QvE verification flows, because it submits `NULL` as the `p_qve_report_info` argument: https://github.com/gramineproject/gramine/blob/4946a0b2224eb989faec3c2e81d8cb3458a9e853/tools/sgx/ra-tls/ra_tls_verify_dcap.c#LL168C33-L168C50

According to the documentation of the DCAP libraries, submitting `NULL` means Untrusted Quote Verification. See for example explanations here: https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/SGX_QuoteVerify_Integration.md

So I don't understand why @IceCY gets these errors. It shouldn't happen in Gramine at all.

**Comment by IceCY** [comment]
Thanks so much for your time. I thought I have identified the reason for the issue now. Previously, I was running my code in the cloud. Recently, I get a SGX server locally. I misunderstood about the PCCS service, and thought that I can just use the cloud provider's PCCS service. In fact, I need to register my machine to Intel Registration Service and maintain a PCCS service locally. I tried this and now I can run the ra-tls-mbedtls successfully.
